magenta/music/musicxml_parser.py

```diff
@@ -0,0 +1,692 @@
+    # Default to one division per measure
+    # From the MusicXML documentation: "The divisions element indicates
+    # how many divisions per quarter note are used to indicate a note's
+    # duration. For example, if duration = 1 and divisions = 2,
+    # this is an eighth note duration."
+    self.divisions = 1
+
+    # Default to a tempo of 120 quarter notes per minute
+    # MusicXML calls this tempo, but Magenta calls this qpm
+    # Therefore, the variable is called qpm, but reads the
+    # MusicXML tempo attribute
+    # (120 qpm is the default tempo according to the
+    # Standard MIDI Files 1.0 Specification)
+    self.qpm = 120
+
+    # Duration of a single quarter note in seconds
+    self.seconds_per_quarter = 0.5
+
+    # Running total of time for the current event in seconds.
+    # Resets to 0 on every part. Affected by <forward> and <backup> elements
+    self.time_position = 0
+
+    # Default to a MIDI velocity of 64 (mf)
+    self.velocity = 64
+
+    # Default MIDI program (1 = grand piano)
+    self.midi_program = 1
+
+    # Current MIDI channel (usually equal to the part number)
+    self.midi_channel = 1
+
+    # Keep track of previous note to get chord timing correct
+    # This variable stores an instance of the Note class (defined below)
+    self.previous_note = None
+
+    # Keep track of current transposition level in +/- semitones.
+    self.transpose = 0
+
+
+class MusicXMLDocument(object):
+  """Internal representation of a MusicXML Document
+
+  Represents the top level object which holds the MusicXML document
+  Responsible for loading the .xml or .mxl file using the get_score method
+  If the file is .mxl, this class uncompresses it
+
+  After the file is loaded, this class then parses the document into memory
+  using the parse method.
+  """
+  def __init__(self, filename):
+    self.filename = filename
+    self.score = self.get_score(filename)
+    self.parts = []
+    self.score_parts = []
+    self.midi_resolution = constants.STANDARD_PPQ
+    self.state = MusicXMLParserState()
+    # Total time in seconds
+    self.total_time_secs = 0
+    self.__parse()
+
+  @staticmethod
+  def get_score(filename):
+    """Given a MusicXML file, return the score as an xml.etree.ElementTree
+
+    Given a MusicXML file, return the score as an xml.etree.ElementTree
+    If the file is compress (ends in .mxl), uncompress it first
+
+    Args:
+        filename: The path of a MusicXML file
+    """
+    if filename.endswith('.mxl'):
+      # Compressed MXL file. Uncompress in memory.
+      filename = ZipFile(filename)
+
+      # Ignore files in the META-INF directory.
+      # Only retrieve the first file
+      namelist = filename.namelist()
+      files = [name for name in namelist if not name.startswith('META-INF/')]
+      compressed_file_name = files[0]
+      score = ET.fromstring(filename.read(compressed_file_name))
+    else:
+      # Uncompressed XML file.
+      tree = ET.parse(filename)
```

Comment:
The call to `parse` should be surrounded by a try/except because it can raise `xml.etree.ElementTree.ParseError`. I noticed this while trying to parse the testdata directory. The `convert_musicxml_dir_to_note_sequences` script is able to catch `MusicXMLConversionError` exceptions and continue, but the `ParseError` exception caused the script to stop because it wasn't caught by anything.